### ISSUE TITLE
Change how Configs are merged

### ISSUE THREAD

**Issue by christetreault** [ucsd-progsys/liquidhaskell#431]
Currently, the Liquid Haskell `Config` type has a `Monoid` instance and this is used to merge configurations (environment variable + command linea args = `Config`). 

After [some discussion](https://github.com/ucsd-progsys/liquid-fixpoint/issues/90#issuecomment-133581372) it seems that this might not be the best way. @gridaphobe suggested a new function `mergeConfig :: Config -> Config -> Config`, not bound by the `Monoid` laws that can be used to merge the configs an a sane manner. This function would merge the configs such that higher priority items override lower priority with `priority = env < file < cmdline`.

**Comment by gridaphobe** [comment]
That's not quite what I meant, so let me elaborate. My suggestion is that instead of having the config-parser return a full `Config` it should return a function from `Config -> Config`, e.g.

```haskell
parseFlag :: String -> (Config -> Config)
-- parseFlag "--no-termination" = \cfg -> cfg { noTermination = True }
-- parseFlag "--idirs src" = \cfg -> cfg { idirs = "src" : idirs cfg }
-- etc.
```

The comments are examples of how this function should behave, not a suggested implementation :) Then we could run the parser on the three different sources of flags and combine the resulting functions (incidentally, these `Config -> Config` functions *do* form a `Monoid`, where `mappend = (.)` and `mempty = id`), giving us a single `Config -> Config` function that we can apply to the default `Config`. So the top-level parser would look something like

```haskell
getOpts :: FilePath -> IO Config
getOpts file = do
  envT <- parseFlags <$> lookupEnv "LIQUIDHASKELL_OPTS"
  fileT <- parseFlags <$> readPragmas file
  argsT <- parseFlags <$> getArgs
  return (argsT . fileT . envT $ defaultConfig)
```

I don't know how difficult it would be to rewrite our config-parser in this manner, it probably depends on the API the underlying parser library provides. Currently we use http://hackage.haskell.org/package/cmdargs.

**Comment by ranjitjhala** [comment]
This seems too complicated.
Can someone remind me of what the problem is with chris' monoid suggestion?

**Comment by christetreault** [comment]
My proposal is to get rid of the Monoid instance and just have a function that combines Configs where one takes precedence. This way we can not have to worry about being good people and [obey the law](https://hackage.haskell.org/package/base-4.8.1.0/docs/Data-Monoid.html)

That said, now that I read the monoid laws, I guess there's no real reason we can't obey them and still do what I suggest...

**Comment by spinda** [comment]
Unless you're going to add `priority` as an actual field of `Config`, I don't think it's possible to follow both of these laws and have prioritization behavior:

    mappend mempty x = x
    mappend x mempty = x

Simpler to just replace the `Monoid` instance with a pair of `defaultConfig` and `mergeConfig` functions.

**Comment by gridaphobe** [comment]
The problem with merging `Config`s (whether or not we use a `Monoid` instance) is that each source of `Config`s specifies a **partial** `Config`. Suppose I have the following file `foo.hs`

```haskell
{-@ LIQUID "--no-termination" @-}
module Foo where
...
```

and run liquid on it with

```
$ liquid --termination foo.hs
```

Since cmdline flags usually take precedence over all else, this invocation of liquid should turn **on** the termination checker. However, we currently merge `Config`s with the following `Monoid` instance

```haskell
instance Monoid Config where
  mappend cfg1 cfg2 = Config { ..., notermination = notermination cfg1 || notermination cfg2 }
```

which will turn **off** the termination checker in my example.

This is clearly wrong, and we could fix it as @christetreault suggests by prioritizing one `Config`, e.g.

```haskell
instance Monoid Config where
  mappend cfg1 cfg2 = Config { .., notermination = notermination cfg2 }
```

With this right-biased merging we can `mappend fileCfg cmdlineCfg` and get the expected result, but alas there's still a problem. What if I run liquid as

```
$ liquid --totality foo.hs
```

instead? Since the `Config` parser has to generate a **full** config for each source it has to fill in unspecified options with the defaults; specifically it will set `notermination = False` in the `cmdlineCfg`. But now when we `mappend fileCfg cmdlineCfg` our `--no-termination` flag in the file will be **ignored**!

**Comment by christetreault** [comment]
This must be a solved problem. The command line parser library probably has some way to handle this...

**Comment by gridaphobe** [comment]
Well, I suspect most programs don't have to deal with parsing flags from multiple sources. On the other hand, they certainly have to deal with sequences of flags like `--termination --no-termination` in a single source. So you're right that the library must be able to deal with this kind of scenario, we're just not giving it all the data at the same time!

A potentially really easy way to fix this issue would be to collect **all** of the flags into a single string and pass that to the library, e.g. `parseFlags (envFlags ++ fileFlags ++ cmdlineFlags)`.

**Comment by spinda** [comment]
The `Mode` passed to `cmdArgsRun` has a field `modeValue`, the starting
value that updates are applied to as flags are processed. Set that field to
the previous (lower-priority) `Config` value when calling `cmdArgsRun`.

**Comment by christetreault** [comment]
@gridaphobe I just checked, you can duplicate fields on the command line, and it will take the rightmost:

```
liquid --verbose --cores=2 --cores=3 omnitest.hs
```

this attempts to solve on 3 cores. This strikes me as a bit of a hack though...

@spinda Your idea seems to have promise to me, however the way I read it, this wants the Config to have multiple constructors like:

```
data Config = FileConfig | EnvironmentConfig | CommandLineConfig
```

... with all fields duplicated. Then we could have Command line > File > Env, but we'd have to copy/paste duplicate the config 3 times. See: [hackage](https://hackage.haskell.org/package/cmdargs-0.7/docs/System-Console-CmdArgs-Implicit.html#v:modes)

It also kind of seems that it would read from one source (i.e. the command line), pick a mode and populate it (The description mentions "multiple modes like cabal" which I read as "cabal install" or "cabal sandbox", etc...), which I'm not sure is what we want.

I feel like @gridaphobe's solution might be the easiest to implement.

**Comment by spinda** [comment]
Are you referring to my first or second reply? The second is how to
implement @gridaphobe's solution with `cmdargs`.

**Comment by spinda** [comment]
Here, this is what I mean. From `CmdLine.hs`:

```haskell
  cfg0    <- envCfg
  cfg1    <- mkOpts =<< cmdArgsRun' config
```

Change this to:

```haskell
  cfg  <- envCfg
  cfg' <- mkOpts =<< cmdArgsRun' $
    config { modeValue = (modeValue config) { cmdArgsValue = cfg } }
```

`cmdArgsRun` takes a value of type `Mode (CmdArgs Config)`. This value contains a starting value of `Config` to make changes to. Normally, this would bee the default `Config`. By changing this starting value to the `Config` parsed from the environment variable, we have `cmdArgsRun` make its modifications to the environment config instead of the default config.

**Comment by gridaphobe** [comment]
The right-to-left behavior does seem strange at first (ie shouldn't conflicting flags trigger an error?) but if you think about it in the context of shell scripts and Makefiles, it's really the only sane thing to do.

Scripts commonly have to construct large command that combine flags from different sources, but the flags are just strings from the script's point of view. It would be a horrible duplication of code to embed a flag parser for, eg gcc, into every Makefile. So you just build up a long string of flags with the understanding that the right-most flags take priority, and let the binary sort it all out. Take a look at the gcc invocations from any moderately sized C project sometime, there's TONS of duplication :)

**Comment by christetreault** [comment]
@spinda I think I'm following now. I'll give your suggestion a shot.

**Comment by christetreault** [comment]
@spinda OK, I'm following now. So we get a `Config` from `envConfig`, then pass it to `CmdArgs` to update, rather than letting `CmdArgs` construct a new config. This should solve the defaults overwriting the passed-in arguments issue.

Now I just have to figure out at what point liquidhaskell parses the source files and updates the config with pragmas.

**Comment by gridaphobe** [comment]
@christetreault you're looking for https://github.com/ucsd-progsys/liquidhaskell/blob/master/src/Language/Haskell/Liquid/GhcInterface.hs#L79-L80

In particular, you probably want to tweak the `withPragmas` function (https://github.com/ucsd-progsys/liquidhaskell/blob/master/src/Language/Haskell/Liquid/CmdLine.hs#L250), which updates the `Config` with the file pragmas.

Interestingly, this has to happen **after** we've parsed the command-line flags, since they tell us which file to check.

**Comment by christetreault** [comment]
@gridaphobe you mentioned a [--termination](https://github.com/ucsd-progsys/liquidhaskell/issues/431#issuecomment-134414802) flag in a previous post. Is this a real flag? I tried in while testing my fix and I just get the help message...

**Comment by christetreault** [comment]
I just pushed [a change](https://github.com/ucsd-progsys/liquidhaskell/blob/merge-config/src/Language/Haskell/Liquid/CmdLine.hs) based on @spinda's suggestion that *should* fix this.

I'll test it some more tomorrow and submit a pull request if all looks well.

**Comment by gridaphobe** [comment]
@christetreault looks like it isn't a real flag. I thought that cmdargs automatically recognizes `--no-foo` for every boolean flag `--foo` (which would actually make it `--no-notermination`) but it doesn't seem to.

What does work however is `--notermination={true,false}`.

**Comment by christetreault** [comment]
Well, I thought I had a fix for this, but my fix has the priority file > cmdline > env. Since we [have to parse the file last](https://github.com/ucsd-progsys/liquidhaskell/issues/431#issuecomment-134722189), we can't use @spinda's suggestion to merge the configs. It seems we'll need some function `merge :: Config -> Config -> Config` after all...

**Comment by gridaphobe** [comment]
I might be wrong about the priorities, file > cmdline > env is probably fine.

**Comment by christetreault** [comment]
@gridaphobe If that is the consensus, then my fix is one PR away! It would certainly be a lot easier...

**Comment by gridaphobe** [comment]
If you want, you could do a quick check to see how GHC handles conflicting options from the cmdline and an `{-# OPTIONS_GHC .. #-}` pragma. But I'd be happy to just do the file > cmdline > env priority.

**Comment by gridaphobe** [comment]
Closed by #436
